components/script/dom/messageport.rs

```diff
@@ -0,0 +1,346 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::codegen::Bindings::EventHandlerBinding::EventHandlerNonNull;
+use dom::bindings::codegen::Bindings::MessagePortBinding::{MessagePortMethods, Wrap};
+use dom::bindings::conversions::{is_array_like, iter_array_object, root_from_handlevalue};
+use dom::bindings::error::{Error, ErrorResult};
+use dom::bindings::inheritance::Castable;
+use dom::bindings::refcounted::Trusted;
+use dom::bindings::reflector::{DomObject, reflect_dom_object};
+use dom::bindings::root::DomRoot;
+use dom::bindings::structuredclone::StructuredCloneData;
+use dom::bindings::transferable::Transferable;
+use dom::eventtarget::EventTarget;
+use dom::globalscope::GlobalScope;
+use dom::messageevent::MessageEvent;
+use dom_struct::dom_struct;
+use js::jsapi::{HandleValue, JSContext, JSStructuredCloneReader, MutableHandleObject};
+use js::jsval::UndefinedValue;
+use std::cell::Cell;
+use std::collections::VecDeque;
+use std::os::raw;
+use std::rc::Rc;
+use std::sync::{Arc, Mutex};
+use task_source::TaskSource;
+
+#[derive(MallocSizeOf)]
+struct PortMessageTask {
+    data: Vec<u8>,
+}
+
+#[derive(JSTraceable, MallocSizeOf)]
+pub struct MessagePortInternal {
+    #[ignore_malloc_size_of = "Task sources are hard"]
+    dom_port: Option<Trusted<MessagePort>>,
+    enabled: bool,
+    has_been_shipped: bool,
+    #[ignore_malloc_size_of = "Defined in std"]
+    entangled_port: Option<Arc<Mutex<MessagePortInternal>>>,
+    pending_port_messages: VecDeque<PortMessageTask>,
+}
+
+impl MessagePortInternal {
+    fn new() -> MessagePortInternal {
+        MessagePortInternal {
+            dom_port: None,
+            enabled: false,
+            has_been_shipped: false,
+            entangled_port: None,
+            pending_port_messages: VecDeque::new(),
+        }
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#dom-messageport-postmessage>
+    // Step 7 substeps
+    fn process_pending_port_messages(&mut self) {
+        if let Some(task) = self.pending_port_messages.pop_front() {
+            // Substep 1
+            let final_target_port = self.dom_port.as_ref().unwrap().root();
+
+            // Substep 2
+            let target_global = final_target_port.global();
+
+            // Substep 3-4
+            rooted!(in(target_global.get_cx()) let mut message_clone = UndefinedValue());
+            let deserialize_result = StructuredCloneData::Vector(task.data).read(
+                &target_global,
+                message_clone.handle_mut()
+            );
+            if !deserialize_result {
+                return;
+            }
+
+            // Substep 5
+
+            // Substep 6
+            MessageEvent::dispatch_jsval(
+                final_target_port.upcast(),
+                &target_global,
+                message_clone.handle()
+            );
+        }
+    }
+
+    fn queue_port_message(&self) {
+        let target_port = self.dom_port.as_ref().unwrap().clone();
```

Comment:
This line is VERY wrong, because the target port would most likely be on a different script thread, and rooting this would cause assertions at runtime.